trie/snapshotTrieStorageManager.go

```diff
@@ -1,74 +1,94 @@
-	epoch              uint32
+	pruningStorer
+	tsm   *trieStorageManager
+	epoch uint32
 }
 
 func newSnapshotTrieStorageManager(tsm *trieStorageManager, epoch uint32) (*snapshotTrieStorageManager, error) {
-	storer, ok := tsm.mainStorer.(snapshotPruningStorer)
+	storer, ok := tsm.mainStorer.(pruningStorer)
 	if !ok {
 		return nil, fmt.Errorf("invalid storer type")
 	}
 
 	return &snapshotTrieStorageManager{
-		trieStorageManager: tsm,
-		mainSnapshotStorer: storer,
-		epoch:              epoch,
+		pruningStorer: storer,
+		tsm:           tsm,
+		epoch:         epoch,
 	}, nil
 }
 
-// Get checks all the storers for the given key, and returns it if it is found
-func (stsm *snapshotTrieStorageManager) Get(key []byte) ([]byte, error) {
-	stsm.storageOperationMutex.Lock()
-	defer stsm.storageOperationMutex.Unlock()
+// Put adds the given value to the last epoch storer
+func (stsm *snapshotTrieStorageManager) Put(key []byte, val []byte) error {
+	if stsm.epoch == 0 {
+		return fmt.Errorf("invalid epoch for snapshot put")
+	}
+
+	return stsm.PutInEpochWithoutCache(key, val, stsm.epoch-1)
+}
 
-	if stsm.closed {
-		log.Debug("snapshotTrieStorageManager get context closing", "key", key)
+// GetFromOldEpochsWithoutAddingToCache searches for the key without adding it to cache if it is found. It starts
+// the search for the key from the currentStorer - epochOffset)
+func (stsm *snapshotTrieStorageManager) GetFromOldEpochsWithoutAddingToCache(key []byte, epochOffset int) ([]byte, error) {
+	if stsm.checkIfClosed(fmt.Sprintf("getFromOldEpochsWithoutAddingToCache context closing, key = %s", hex.EncodeToString(key))) {
 		return nil, errors.ErrContextClosing
 	}
 
-	val, err := stsm.mainSnapshotStorer.GetFromOldEpochsWithoutAddingToCache(key)
-	if isClosingError(err) {
-		return nil, err
-	}
-	if len(val) != 0 {
-		return val, nil
+	return stsm.pruningStorer.GetFromOldEpochsWithoutAddingToCache(key, epochOffset)
+}
+
+// PutInEpochWithoutCache saves the data in the given epoch storer
+func (stsm *snapshotTrieStorageManager) PutInEpochWithoutCache(key []byte, data []byte, epoch uint32) error {
+	if stsm.checkIfClosed(fmt.Sprintf("putInEpochWithoutCache context closing, key = %s", hex.EncodeToString(key))) {
+		return errors.ErrContextClosing
 	}
 
-	return stsm.getFromOtherStorers(key)
+	return stsm.pruningStorer.PutInEpochWithoutCache(key, data, epoch)
 }
 
-// Put adds the given value to the main storer
-func (stsm *snapshotTrieStorageManager) Put(key, data []byte) error {
-	stsm.storageOperationMutex.Lock()
-	defer stsm.storageOperationMutex.Unlock()
+// GetFromEpochWithoutCache seeks the key in the specified epoch storer
+func (stsm *snapshotTrieStorageManager) GetFromEpochWithoutCache(key []byte, epoch uint32) ([]byte, error) {
+	if stsm.checkIfClosed(fmt.Sprintf("getFromEpochWithoutCache context closing, key = %s", hex.EncodeToString(key))) {
+		return nil, errors.ErrContextClosing
+	}
+
+	return stsm.pruningStorer.GetFromEpochWithoutCache(key, epoch)
```

Comment:
what happens if we end up here and in exact that time the pruning storer is closed? This is an edgecase that can happen

Valid here and on L72, L54

Reply:
Refactored.